cmd/cli/update.go

```diff
@@ -0,0 +1,170 @@
+*/
+package main
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/openservicemesh/osm/pkg/cli"
+	"github.com/openservicemesh/osm/pkg/constants"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	helm "helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"k8s.io/client-go/kubernetes"
+)
+
+const updateDesc = `
+This command updates an osm control plane on the Kubernetes cluster.
+
+The default Kubernetes namespace that gets updated is called osm-system. 
+To update a control plane components in a different namespace, use the 
+--namespace flag.
+
+Example:
+  $ osm update --namespace hello-world
+
+Multiple control plane installations can exist within a cluster. Each
+control plane is given a cluster-wide unqiue identifier called mesh name.
+A mesh name can be passed in via the --mesh-name flag. By default, the
+mesh-name name will be set to "osm." The mesh name must conform to same
+guidelines as a valid Kubernetes label value. Must be 63 characters or
+less and must be empty or begin and end with an alphanumeric character
+([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and
+alphanumerics between.
+
+Example:
+  $ osm update --mesh-name "hello-osm"
+
+The mesh name is used in various ways like for naming Kubernetes resources as
+well as for adding a Kubernetes Namespace to the list of Namespaces a control
+plane should watch for sidecar injection of Envoy proxies.
+`
+
+type updateCmd struct {
+	out       io.Writer
+	clientSet kubernetes.Interface
+	// action config is used to resolve the helm chart values
+	cfg actionConfig
+}
+
+func newUpdateCmd(config *helm.Configuration, out io.Writer) *cobra.Command {
+	upd := &updateCmd{
+		out: out,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "update",
+		Short: "update osm control plane configuration",
+		Long:  updateDesc,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			kubeconfig, err := settings.RESTClientGetter().ToRESTConfig()
+			if err != nil {
+				return errors.Errorf("Error fetching kubeconfig")
+			}
+
+			clientset, err := kubernetes.NewForConfig(kubeconfig)
+			if err != nil {
+				return errors.Errorf("Could not access Kubernetes cluster. Check kubeconfig")
+			}
+			upd.clientSet = clientset
+			return upd.run(config)
+		},
+	}
+
+	f := cmd.Flags()
+	f.StringVar(&upd.cfg.containerRegistry, "container-registry", "openservicemesh", "container registry that hosts control plane component images")
+	f.StringVar(&upd.cfg.osmImageTag, "osm-image-tag", "v0.2.0", "osm image tag")
+	f.StringVar(&upd.cfg.containerRegistrySecret, "container-registry-secret", "acr-creds", "name of Kubernetes secret for container registry credentials to be created if it doesn't already exist")
+	f.StringVar(&upd.cfg.chartPath, "osm-chart-path", "", "path to osm chart to override default chart")
+	f.StringVar(&upd.cfg.certManager, "certificate-manager", defaultCertManager, "certificate manager to use (tresor or vault)")
```

Comment:
We are not ready to support some of these flags in the update action. The only options that should be able to be configured are the ones defined in the osm configmap in charts/osm/templates/osm-configmap.yaml